File: android/app/src/main/kotlin/com/tbtrapp/calls/CallConnectionMonitor.kt
package com.tbtrapp.calls

import android.os.Handler
import android.os.Looper
import android.util.Log
import org.webrtc.PeerConnection

/**
 * Independent safety net. If the peer connection never reaches
 * CONNECTED within [timeoutMs], [onTimeout] is fired.
 * This prevents getting stuck forever on "Connecting…" when the
 * caller crashes, TURN is broken, or Firebase signaling races.
 */
class CallConnectionMonitor(
    private val timeoutMs: Long = 30000L,
    private val onTimeout: () -> Unit
) {
    private val handler = Handler(Looper.getMainLooper())
    @Volatile private var isConnected = false
    @Volatile private var isCancelled = false

    private val timeoutRunnable = Runnable {
        if (!isConnected && !isCancelled) {
            Log.w("CallConnectionMonitor", "Connection timeout after ${timeoutMs}ms — forcing hangup")
            onTimeout()
        }
    }

    /** Start the timer. Call this when the call UI opens. */
    fun start() {
        isConnected = false
        isCancelled = false
        handler.postDelayed(timeoutRunnable, timeoutMs)
    }

    /** Call from PeerConnection.Observer.onConnectionStateChange when CONNECTED. */
    fun markConnected() {
        if (isConnected) return
        isConnected = true
        handler.removeCallbacks(timeoutRunnable)
    }

    /** Call this in onDestroy() / hangUp() to avoid leaking the handler. */
    fun cancel() {
        isCancelled = true
        handler.removeCallbacks(timeoutRunnable)
    }

    /** Convenience wrapper. */
    fun onPeerConnectionStateChange(state: PeerConnection.PeerConnectionState?) {
        when (state) {
            PeerConnection.PeerConnectionState.CONNECTED -> markConnected()
            else -> { /* wait */ }
        }
    }
}

File: lib/calls/call_connection_monitor.dart
import 'dart:async';
import 'package:flutter/foundation.dart';
import 'package:flutter_webrtc/flutter_webrtc.dart';

class CallConnectionMonitor {
  final int timeoutMs;
  final VoidCallback onTimeout;

  Timer? _timer;
  bool _isConnected = false;
  bool _isCancelled = false;

  CallConnectionMonitor({
    this.timeoutMs = 30000,
    required this.onTimeout,
  });

  void start() {
    _isConnected = false;
    _isCancelled = false;
    _timer?.cancel();
    _timer = Timer(Duration(milliseconds: timeoutMs), () {
      if (!_isConnected && !_isCancelled) {
        debugPrint('CallConnectionMonitor: Connection timeout after ${timeoutMs}ms — forcing hangup');
        onTimeout();
      }
    });
  }

  void markConnected() {
    if (_isConnected) return;
    _isConnected = true;
    _timer?.cancel();
  }

  void cancel() {
    _isCancelled = true;
    _timer?.cancel();
  }

  void onPeerConnectionStateChange(RTCPeerConnectionState? state) {
    if (state == RTCPeerConnectionState.RTCPeerConnectionStateConnected) {
      markConnected();
    }
  }
}